state/factory/minter.go

```
@@ -7,44 +7,130 @@ package factory
+	ap            actpool.ActPool
+	timeout       time.Duration
+	blockPreparer *blockPreparer
+	proposalPool  *proposalPool
+	mu            sync.Mutex
 }
 
 // NewMinter creates a wrapper instance
-func NewMinter(f Factory, ap actpool.ActPool, opts ...MintOption) blockchain.BlockBuilderFactory {
-	m := &minter{f: f, ap: ap}
+func NewMinter(f Factory, ap actpool.ActPool, opts ...MintOption) *Minter {
+	m := &Minter{
+		f:             f,
+		ap:            ap,
+		blockPreparer: newBlockPreparer(),
+		proposalPool:  newProposalPool(),
+	}
 	for _, opt := range opts {
 		opt(m)
 	}
 	return m
 }
 
-// NewBlockBuilder implements the BlockMinter interface
-func (m *minter) NewBlockBuilder(ctx context.Context, sign func(action.Envelope) (*action.SealedEnvelope, error)) (*block.Builder, error) {
+func (m *Minter) Init(root hash.Hash256) {
+	m.proposalPool.Init(root)
+}
+
+func (m *Minter) Mint(ctx context.Context, pk crypto.PrivateKey) (*block.Block, error) {
+	bcCtx := protocol.MustGetBlockchainCtx(ctx)
+	blkCtx := protocol.MustGetBlockCtx(ctx)
+
+	// create a new block
+	blk, err := m.blockPreparer.PrepareOrWait(ctx, bcCtx.Tip.Hash[:], blkCtx.BlockTimeStamp, func() (*block.Block, error) {
+		blk, err := m.mint(ctx, pk)
+		if err != nil {
+			return nil, err
+		}
+		if err = m.proposalPool.AddBlock(blk); err != nil {
+			log.L().Error("failed to add block to proposal pool", zap.Error(err))
+		}
+		return blk, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return blk, nil
+}
+
+func (m *Minter) AddProposal(blk *block.Block) error {
+	return m.proposalPool.AddBlock(blk)
+}
+
+func (m *Minter) ReceiveBlock(blk *block.Block) error {
+	prevHash := blk.PrevHash()
+	l := log.L().With(zap.Uint64("height", blk.Height()), log.Hex("prevHash", prevHash[:]), zap.Time("timestamp", blk.Timestamp()))
+	if err := m.blockPreparer.ReceiveBlock(blk); err != nil {
+		l.Error("failed to receive block", zap.Error(err))
+	}
+	if err := m.proposalPool.ReceiveBlock(blk); err != nil {
```

the error is ignored? will it cause a problem?